### PR TITLE
WIP: [IE CORE]  enable plugins & dependent libs loading using absolute path

### PR DESCRIPTION
**Pull request by esmirno**
not sure about extra safety, but please give feed back about general idea
Currently this allowed to use plugins.xml file to specify full path to specific plugin with it's all dependency, not to be persisted in CWD or in PATH. Since that requires physical access to settings xml, i doubt we broke something.

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
@esmirno can we hide all implementation details in cpp files? In general I want to hide linux and windows folder while public API is OS-indepenent.

**Comment by esmirno** [comment]
@ilya-lavrenov  agree - so best approach would be to move all OS - stuff into src folder and use only Core API. But i afraid it is significant refactoring and can be done separately. Regarding this PR not sure whether it worth stripping it - into include/src folder.

**Comment by ilya-lavrenov** [comment]
@esmirno I don't think it's a bug refactoring. The API for both classes is the same, could you please try to hide implementations? Because with your changes more and more implementation details are in header files.

**Comment by esmirno** [comment]
@ilya-lavrenov ok i'll try, thanks

**Comment by ilya-lavrenov** [comment]
> @ilya-lavrenov ok i'll try, thanks

Do you have any progress?

**Comment by ilya-lavrenov** [comment]
Trying to move implementation details to cpp files in https://github.com/openvinotoolkit/openvino/pull/1556
Will notify once it's ready for review.

File: inference-engine/include/details/os/win_shared_object_loader.h [hunk 1]
```diff
@@ -62,8 +124,11 @@ class SharedObjectLoader {
      */
     explicit SharedObjectLoader(LPCWSTR pluginName) {
         ExcludeCurrentDirectory();
+        auto oldDir = IncludePluginDirectoryW(pluginName);
 
         shared_object = LoadLibraryW(pluginName);
+
+        SetDllDirectoryW(oldDir.c_str());
```
Comment (asomsiko) [inline inference-engine/include/details/os/win_shared_object_loader.h:131]: Is the intent was to roll back changes? SetDllDirectory adds, not replace.

**Comment by asomsiko** [comment]
I see no issues if the plugin can control search path from plugins.xml. Requirements to load from trusted locations are meet.

File: inference-engine/include/details/os/win_shared_object_loader.h [hunk 2]
```diff
@@ -62,8 +124,11 @@ class SharedObjectLoader {
      */
     explicit SharedObjectLoader(LPCWSTR pluginName) {
         ExcludeCurrentDirectory();
+        auto oldDir = IncludePluginDirectoryW(pluginName);
```
Comment (asomsiko) [inline inference-engine/include/details/os/win_shared_object_loader.h:127]: ExcludeCurrentDirectory calls SetDllDirectoryW(""). Could you debug and confirm IncludePluginDirectoryW returns "" (not nullptr) such that consequent SetDllDirectoryW will have the effect of ExcludeCurrentDirectory?

**Comment by ilya-lavrenov** [comment]
@esmirno please, finalize this and resolve merge conflicts. After we can merge this

**Comment by ilya-lavrenov** [comment]
@esmirno ping

**Comment by esmirno** [comment]
@ilya-lavrenov  probably can do on next week. OOO till that. thanks, for remind

**Comment by ilya-lavrenov** [comment]
> @ilya-lavrenov probably can do on next week. OOO till that. thanks, for remind

ok, because I want to include this in 2021.1

File: inference-engine/include/details/os/win_shared_object_loader.h [hunk 3]
```diff
@@ -48,6 +48,68 @@ class SharedObjectLoader {
...
+            return path;
+        }
+        std::basic_string<CHAR> original(path);
+        return original.substr(0, std::distance(path, pos));
+    }
+    std::basic_string<CHAR> IncludePluginDirectoryA(LPCSTR path) {
+        std::vector<CHAR> lpBuffer;
+        DWORD nBufferLength;
+
+        nBufferLength = GetDllDirectoryW(0, nullptr);
```
[the remark this answers is not in the document]
Comment (esmirno) [reply inference-engine/include/details/os/win_shared_object_loader.h]: done

**Review by ilya-lavrenov — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
@esmirno  please, make sure CI is green

**Comment by ilya-lavrenov** [comment]
@esmirno please, make sure CI is green

**Comment by ilya-lavrenov** [comment]
Replaced by https://github.com/openvinotoolkit/openvino/pull/3639

**Pull request closed without merging**